### ISSUE TITLE
rate limit requested quota is not configured

### ISSUE THREAD

**Issue by rookiefiend** [istio/istio#14658]
**Bug description**
  when I perform [enabling rate limits  task](https://istio.io/docs/tasks/policy-enforcement/rate-limiting/#conditional-rate-limits). Wether I use memory quota or redis quota, the rate limits is not work, and I can find warn in mixer --- `warn	Requested quota 'requestcountquota' is not configured`
**Affected product area (please put an X in all that apply)**
  istio 1.1.6
[ ] Configuration Infrastructure
  I use istio in kubernetes.
[ ] Docs
  
[ ] Installation
  I install istio by helm, use configuation that disable prometheus base on default configuration.
[ ] Networking
  calico
[ ] Performance and Scalability
  request: cpu: 500m, memory: 500M
[ ] Policies and Telemetry
[ ] Security
   not use
[ ] Test and Release
   release istio 1.1.6
[ ] User Experience
  
**Expected behavior**
  I want the rate limit can work.
**Steps to reproduce the bug**
  I don't sure other will meet this problem, I Install istio 1.1.6 with helm that default configuartion, note I disabel the prometheus. then  follow [enabling rate limits task](https://istio.io/docs/tasks/policy-enforcement/rate-limiting/#conditional-rate-limits) step by step.
**Version (include the output of `istioctl version --remote` and `kubectl version`)**
  - istioctl version --remote
 ```
client version: version.BuildInfo{Version:"1.1.6", GitRevision:"04850e14d38a69a38c16c800e237b1108056513e", User:"root", Host:"1844d064-72cc-11e9-a0d5-0a580a2c0304", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.5-11-g04850e1"}
citadel version: version.BuildInfo{Version:"1.1.6", GitRevision:"04850e14d38a69a38c16c800e237b1108056513e-dirty", User:"root", Host:"1844d064-72cc-11e9-a0d5-0a580a2c0304", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.5-11-g04850e1"}
galley version: version.BuildInfo{Version:"1.1.6", GitRevision:"04850e14d38a69a38c16c800e237b1108056513e-dirty", User:"root", Host:"1844d064-72cc-11e9-a0d5-0a580a2c0304", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.5-11-g04850e1"}
ingressgateway version: version.BuildInfo{Version:"1.1.6", GitRevision:"04850e14d38a69a38c16c800e237b1108056513e", User:"root", Host:"1844d064-72cc-11e9-a0d5-0a580a2c0304", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.5-11-g04850e1"}
pilot version: version.BuildInfo{Version:"1.1.6", GitRevision:"04850e14d38a69a38c16c800e237b1108056513e-dirty", User:"root", Host:"1844d064-72cc-11e9-a0d5-0a580a2c0304", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.5-11-g04850e1"}
policy version: version.BuildInfo{Version:"1.1.6", GitRevision:"04850e14d38a69a38c16c800e237b1108056513e-dirty", User:"root", Host:"1844d064-72cc-11e9-a0d5-0a580a2c0304", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.5-11-g04850e1"}
sidecar-injector version: version.BuildInfo{Version:"1.1.6", GitRevision:"04850e14d38a69a38c16c800e237b1108056513e-dirty", User:"root", Host:"1844d064-72cc-11e9-a0d5-0a580a2c0304", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.5-11-g04850e1"}
telemetry version: version.BuildInfo{Version:"1.1.6", GitRevision:"04850e14d38a69a38c16c800e237b1108056513e-dirty", User:"root", Host:"1844d064-72cc-11e9-a0d5-0a580a2c0304", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.5-11-g04850e1"}
 ```
- kubectl version
```
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.5", GitCommit:"51dd616cdd25d6ee22c83a858773b607328a18ec", GitTreeState:"archive", BuildDate:"2019-01-25T08:57:03Z", GoVersion:"go1.10.7", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.5", GitCommit:"51dd616cdd25d6ee22c83a858773b607328a18ec", GitTreeState:"archive", BuildDate:"2019-01-25T08:57:03Z", GoVersion:"go1.10.7", Compiler:"gc", Platform:"linux/amd64"}
```
**How was Istio installed?**
  use helm that disable prometheus base on default configuration.
**Environment where bug was observed (cloud vendor, OS, etc)**
   private cloud
Additionally, please consider attaching a [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) by attaching
the dump file to this issue.

**Comment by gargnupur** [comment]
Can you share your mixer logs along with the config you are using?

**Comment by rookiefiend** [comment]
- policy mixer log
```
[root@master1 ~]# kubectl -n istio-system logs -f istio-policy-5bcf878994-k5grw -c mixer --tail=0
scvg461: 0 MB released
scvg461: inuse: 32, idle: 10, sys: 43, released: 8, consumed: 35 (MB)
2019-06-11T01:22:11.055667Z	warn	Requested quota 'requestcountquota' is not configured
2019-06-11T01:22:11.831993Z	warn	Requested quota 'requestcountquota' is not configured
2019-06-11T01:22:12.400319Z	warn	Requested quota 'requestcountquota' is not configured
```
the log print when I access the productpage service
- redis quota configuration
```
apiVersion: config.istio.io/v1alpha2
kind: handler
metadata:
  name: redishandler
  namespace: istio-system
spec:
  compiledAdapter: redisquota
  params:
    redisServerUrl: shen-space-redis-master.shen-space:6379
    connectionPoolSize: 10
    quotas:
    - name: requestcountquota.instance.istio-system
      maxAmount: 500
      validDuration: 1s
      bucketDuration: 500ms
      rateLimitAlgorithm: ROLLING_WINDOW
      # The first matching override is applied.
      # A requestcount instance is checked against override dimensions.
      overrides:
      # The following override applies to 'reviews' regardless
      # of the source.
      - dimensions:
          destination: reviews
        maxAmount: 1
      # The following override applies to 'productpage' when
      # the source is a specific ip address.
      - dimensions:
          destination: productpage
          source: "10.28.11.20"
        maxAmount: 500
      # The following override applies to 'productpage' regardless
      # of the source.
      - dimensions:
          destination: productpage
        maxAmount: 2
---
apiVersion: config.istio.io/v1alpha2
kind: instance
metadata:
  name: requestcountquota
  namespace: istio-system
spec:
  compiledTemplate: quota
  params:
    dimensions:
      source: request.headers["x-forwarded-for"] | "unknown"
      destination: destination.labels["app"] | destination.workload.name | "unknown"
      destinationVersion: destination.labels["version"] | "unknown"
---
apiVersion: config.istio.io/v1alpha2
kind: QuotaSpec
metadata:
  name: request-count
  namespace: istio-system
spec:
  rules:
  - quotas:
    - charge: 1
      quota: requestcountquota
---
apiVersion: config.istio.io/v1alpha2
kind: QuotaSpecBinding
metadata:
  name: request-count
  namespace: istio-system
spec:
  quotaSpecs:
  - name: request-count
    namespace: istio-system
  services:
  - name: productpage
    namespace: default
    #  - service: '*'  # Uncomment this to bind *all* services to request-count
---
apiVersion: config.istio.io/v1alpha2
kind: rule
metadata:
  name: quota
  namespace: istio-system
spec:
  # quota only applies if you are not logged in.
  # match: match(request.headers["cookie"], "session=*") == false
  actions:
  - handler: redishandler
    instances:
    - requestcountquota
---
```

**Comment by rookiefiend** [comment]
> Can you share your mixer logs along with the config you are using?

I use the istio offical yaml that only change the redis url, and I also meet this question when use memory quota.
Thanks.
@gargnupur

**Comment by chenyb4** [comment]
@rookiefiend You need restart istio-pilot and istio-policy pod in istio-system

**Comment by rookiefiend** [comment]
@chenyb4 Thanks. when I restart istio-policy and istio-pilot, I try use memory quota success, but user redist qutoa still has that question. Is it possible redis problem?

**Comment by gargnupur** [comment]
@rookiefiend : Couple of questions:
1) Can you enable debug logs for your deployment?
2) Can you share config dump of productpage? Command to do that would be something like this: kubectl  exec -i -t <your producptapepod> -c istio-proxy -- curl http://localhost:15000/config_dump > productpage.json

**Comment by gargnupur** [comment]
Also, which version of Istio are you using?

**Comment by chenyb4** [comment]
In my test, I don't use redis quota

Nupur Garg <notifications@github.com> 于2019年6月27日周四 上午4:10写道：

**Comment by jstew7** [comment]
I'm seeing the same issue with redisquota. It works without any overrides, but as soon as I add an override to the handler it throws that warning and does not honor the quota. If I switch to memquota it works fine with same settings. I'm on Istio 1.2.0.

**Comment by gargnupur** [comment]
I just tested on Istio 1.2.0 and I am seeing the same thing... Looking why this is happening

/cc @douglas-reid @mandarjog @kyessenov

**Comment by gargnupur** [comment]
Sorry red herring.. I saw it because I set the wrong redis url and had error like this in my logs:
"2019-07-02T18:48:37.610246Z	error	Unable to initialize adapter: snapshot='4', handler='redishandler.istio-system', adapter='redisquota', err='adapter builder validation failed: adapter validation failed : builder validation failed: '1 error occurred:
	* redisquota: could not create a connection to redis server: dial tcp: lookup redis-release-master.istio.redis on 10.35.240.10:53: no such host

''.
Please remove the handler or fix the configuration.
2019-07-02T18:48:37.611789Z	warn	Unable to find a handler for action. rule[action]='quota.rule.istio-system[0]', handler='redishandler.istio-system'"

Verified that it's working even with overrides.
@jstew7 : Can you share your full logs?

**Comment by jstew7** [comment]
@gargnupur : I just figured out what I had done wrong. When I switched from memquota to redisquota I had left a "validDuration" param in the override dimension. I didn't realize that was only part of memquota and had missed that subtle difference. Once I removed it is working as expected. 

If I could've gotten more debug logs it might've been easier to figure out (I tried turning on all the debug flags I could find). The only message I saw was the one warning.
